### PR TITLE
Metricset/kafka consumer groups

### PR DESCRIPTION
**Pull request by urso**
Requires #3239

### CODE REVIEW

File: metricbeat/module/kafka/broker.go [hunk 1]
```diff
@@ -140,6 +161,82 @@ func (b *Broker) PartitionOffset(
...
+	}
+	return groups, nil
+}
+
+// DescribeGroups fetches group details from broker.
+func (b *Broker) DescribeGroups(
+	queryGroups []string,
+) (map[string]GroupDescription, error) {
+	requ := &sarama.DescribeGroupsRequest{Groups: queryGroups}
+	resp, err := b.b.DescribeGroups(requ)
```
Comment (ruflin) [inline metricbeat/module/kafka/broker.go]: b.b. is not very readable ;-)

File: metricbeat/module/kafka/consumergroup/_meta/data.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    },
+    "metricset":{
+        "host":"localhost",
+        "module":"mysql",
+        "name":"status",
+        "rtt":44269
+    },
+    "kafka":{
+        "consumergroup":{
+            "example": "consumergroup"
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/_meta/data.json]: We should populate this for the docs.

File: metricbeat/module/kafka/consumergroup/_meta/fields.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+          type: long
+          description: >
+            Broker id
+
+        - name: address
+          type: keyword
+          description: >
+            Broker address
+
+    - name: group
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/_meta/fields.yml]: This should be `id`, as it will be `kafka.consumegroup.id`. Alternative is `kafka.conumser.group`. Is the partition, topic, offset specific to a consumergroup or a consumer?

File: metricbeat/module/kafka/consumergroup/config.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	TLS      *outputs.TLSConfig `config:"ssl"`
+	Username string             `config:"username"`
+	Password string             `config:"password"`
+	ClientID string             `config:"client_id"`
+
+	Groups []string `config:"groups"`
+	Topics []string `config:"topics"`
+}
+
+var defaultConfig = metricsetConfig{
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/config.go]: We should move this to the kafka module later as I think it is the same for both metricsets.

File: metricbeat/module/kafka/consumergroup/consumergroup.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
+package consumergroup
+
+import (
+	"crypto/tls"
+
+	"github.com/Shopify/sarama"
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/consumergroup.go]: missing newline

File: metricbeat/module/kafka/consumergroup/consumergroup.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	groups nameSet
+}
+
+type groupAssignment struct {
+	clientID   string
+	memberID   string
+	clientHost string
+}
+
+var debugf = logp.MakeDebug("kafka")
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/consumergroup.go]: kafka.consumergroup?

File: metricbeat/module/kafka/consumergroup/mock_test.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
+package consumergroup
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/Shopify/sarama"
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/mock_test.go]: newline

File: metricbeat/module/kafka/consumergroup/query.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+	if len(groups) == 0 {
+		return nil
+	}
+
+	debugf("known consumer groups: ", groups)
+
+	wg := sync.WaitGroup{}
+	results := make(chan result, len(groups))
+	for _, group := range groups {
+		group := group
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/query.go]: copy needed?

File: metricbeat/module/kafka/consumergroup/query.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	resp, err := b.DescribeGroups(groupIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	groups := map[string]map[string]map[int32]groupAssignment{}
+
+groupLoop:
+	for groupID, info := range resp {
+		G := groups[groupID]
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/query.go]: upper case?

File: metricbeat/module/kafka/consumergroup/query.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+			}
+
+			meta := groupAssignment{
+				memberID:   memberID,
+				clientID:   clientID,
+				clientHost: clientHost,
+			}
+
+			for topic, partitions := range memberDescr.Topics {
+				T := G[topic]
```
Comment (ruflin) [inline metricbeat/module/kafka/consumergroup/query.go]: upper case?

File: metricbeat/module/kafka/consumergroup/_meta/fields.yml (hunk 3 above)
Comment (urso) [inline metricbeat/module/kafka/consumergroup/_meta/fields.yml]: will rename it to `id`

File: metricbeat/module/kafka/consumergroup/consumergroup.go (hunk 6 above)
Comment (urso) [inline metricbeat/module/kafka/consumergroup/consumergroup.go]: partitions is using `kafka` as well. we should clean this in a followup PR. Do we follow a debug selector naming standard in metricbeat?

File: metricbeat/module/kafka/consumergroup/query.go (hunk 8 above)
Comment (urso) [inline metricbeat/module/kafka/consumergroup/query.go]: Yes, as `group` is used inside the go-routine. Copy is used to capture the `group` in the local scope.

Alternatively one can write:

```
for group := range groups {
  go func(group string) {
   ...
  }(group)
}
```

which is similar to writing:

```
for group := range groups {
  group := group
  go func() {
   ...
  }()
}
```

File: metricbeat/module/kafka/consumergroup/query.go (hunk 9 above)
Comment (urso) [inline metricbeat/module/kafka/consumergroup/query.go]: I tend to use uppercase for local sets by convention (e.g. like N,M,L for length). Here G for group and T for topic. So not to shadow (overuse) `topic` and `group` used for correlating information. The 'sets' G and T hold results, so using a single upper-case makes them kinda stick out in this 'heavy' (potentially hard to follow) function.
